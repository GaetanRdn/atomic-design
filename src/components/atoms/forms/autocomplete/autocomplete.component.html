<input adrInput [value]="displayedValue" (focus)="openPanel()" (valueChange)="filterOptions($event)" [disabled]="disabled" />
<div *ngIf="isOpen" class="adr-autocomplete-options">
    <span class="adr-option" [class.adr-selected]="isSelected(option)" *ngFor="let option of displayedValues$ | async; trackBy:
    trackByFn"
          (click)="select(option)">{{ displayOptionFn(option) }}</span>
</div>
