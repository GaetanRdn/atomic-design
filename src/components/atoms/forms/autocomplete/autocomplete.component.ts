import {
  ChangeDetectionStrategy,
  Component,
  ElementRef,
  EventEmitter,
  HostListener,
  Input,
  Output,
} from "@angular/core";
import { BehaviorSubject, Observable } from "rxjs";
import {
  DisplayFn,
  OpenOn,
} from "src/components/atoms/forms/autocomplete/autocomplete.models";
import { CoerceBoolean } from "src/components/core/common/coerce-boolean-inputs.decorator";

@Component({
  selector: "adr-autocomplete",
  host: {
    "[class.adr-opened]": "isOpen",
    "[attr.required]": "required || null",
  },
  templateUrl: "./autocomplete.component.html",
  styleUrls: ["./autocomplete.component.scss"],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class AutocompleteComponent<T> {
  @Input()
  set options(options: T[]) {
    this._options = (options || []).sort((a: T, b: T) =>
      this.displayOptionFn(a) < this.displayOptionFn(b) ? -1 : 1
    );

    this._displayedValues$.next(this._options);
  }

  private _options: T[] = [];

  @Input()
  public value: T | null = null;

  @Input()
  @CoerceBoolean()
  public required: boolean = false;

  @Input()
  public openOn: OpenOn = "focus";

  @Output()
  public readonly valueChange: EventEmitter<T | null> = new EventEmitter<T | null>();

  private _displayedValues$: BehaviorSubject<T[]> = new BehaviorSubject<T[]>(
    this.options
  );

  public displayedValues$: Observable<
    T[]
  > = this._displayedValues$.asObservable();

  get displayedValue(): string {
    return this.value !== null ? this.displayOptionFn(this.value) : "";
  }

  private _isOpen: boolean = false;

  get isOpen(): boolean {
    return this._isOpen;
  }

  constructor(private _elementRef: ElementRef) {}

  @Input()
  public displayOptionFn: DisplayFn<T> = (option: T): string =>
    (option as unknown) as string;

  @HostListener("document:click", ["$event.target"])
  public onClick(event: EventTarget): void {
    if (!this._elementRef.nativeElement.contains(event)) {
      this._isOpen = false;
    }
  }

  public openPanel(): void {
    if (this.openOn === "focus") {
      this._isOpen = true;
    }
  }

  public select(option: T): void {
    let hasChanged = false;

    if (this.value !== option) {
      this.value = option;
      hasChanged = true;
    } else if (!this.required) {
      this.value = null;
      hasChanged = true;
    }

    if (hasChanged) {
      this._isOpen = false;
      this.valueChange.emit(this.value);
    }
  }

  public filterOptions(inputValue: string): void {
    if (this.openOn === "input" && !this.isOpen) {
      this._isOpen = true;
    }
    this._displayedValues$.next(
      this._options.filter((option: T) =>
        this.displayOptionFn(option)
          .toLocaleLowerCase()
          .includes(inputValue.toLocaleLowerCase())
      )
    );
  }
}
