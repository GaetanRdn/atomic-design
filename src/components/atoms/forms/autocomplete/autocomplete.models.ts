export type DisplayFn<T> = (option: T) => string;

export type OpenOn = "focus" | "input";
