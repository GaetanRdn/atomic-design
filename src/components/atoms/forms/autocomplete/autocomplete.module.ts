import { CommonModule } from "@angular/common";
import { NgModule } from "@angular/core";
import { InputModule } from "src/components/atoms/forms/input/input.module";
import { AutocompleteComponent } from "./autocomplete.component";

@NgModule({
  declarations: [AutocompleteComponent],
  exports: [AutocompleteComponent],
  imports: [CommonModule, InputModule],
})
export class AutocompleteModule {}
