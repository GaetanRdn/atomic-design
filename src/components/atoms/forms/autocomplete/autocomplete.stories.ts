import { FormControl, ReactiveFormsModule } from '@angular/forms';
import { action } from '@storybook/addon-actions';
import { Meta, Story } from '@storybook/angular';
import { AutocompleteComponent } from './autocomplete.component';
import { AutocompleteModule } from './autocomplete.module';

export default {
    title: 'atoms/forms/autocomplete',
    component: AutocompleteComponent,
} as Meta<AutocompleteComponent<string>>;

const template: Story<AutocompleteComponent<string>> = (args: AutocompleteComponent<string>) => ({
    moduleMetadata: {
        imports: [AutocompleteModule],
    },
    props: {
        ...args,
        valueChange: action('valueChange'),
    },
    template: `
  <adr-autocomplete [value]="value"
  [openOn]="openOn"
  [options]="options" 
  [required]="required" 
  [disabled]="disabled" 
  (valueChange)="valueChange($event)">
  </adr-autocomplete>`,
});

export const basic = template.bind({});
basic.args = {
    options: ['Gaetan', 'Soren', 'Bernard'],
    value: '',
};

interface Person {
    id?: number;
    firstName: string;
    name: string;
}

const objectsValuesTemplate: Story<AutocompleteComponent<Person>> = (
    args: AutocompleteComponent<Person>,
) => ({
    moduleMetadata: {
        imports: [AutocompleteModule],
    },
    props: {
        ...args,
        valueChange: action('valueChange'),
        displayOptionFn: (option: Person): string => `${option.firstName} - ${option.name}`,
    },
    template: `
  <adr-autocomplete [value]="value"
  [openOn]="openOn"
  [options]="options" 
  [required]="required" 
  [disabled]="disabled"
  (valueChange)="valueChange($event)"
  [displayOptionFn]="displayOptionFn">
  </adr-autocomplete>`,
});

export const complexValues = objectsValuesTemplate.bind({});
complexValues.args = {
    options: [
        { id: 1, firstName: 'Gaetan', name: 'Redin' },
        { id: 2, firstName: 'Soren', name: 'Redin' },
        { id: 3, firstName: 'Lord', name: 'Voldemor' },
    ],
    value: { id: 2, firstName: 'Soren', name: 'Redin' },
};

const reactiveFormTemplate: Story<AutocompleteComponent<Person>> = (
    args: AutocompleteComponent<Person>,
) => ({
    moduleMetadata: {
        imports: [AutocompleteModule, ReactiveFormsModule],
    },
    props: {
        ...args,
        valueChange: action('valueChange'),
        value: null,
        displayOptionFn: (option: Person): string => `${option.firstName} - ${option.name}`,
        control: new FormControl({ id: 2, firstName: 'Soren', name: 'Redin' }),
        identityFn: (p: Person): any => p.id,
    },
    template: `
  <adr-autocomplete 
  [formControl]="control"
  [openOn]="openOn"
  [options]="options" 
  [required]="required" 
  (valueChange)="valueChange($event)"
  [displayOptionFn]="displayOptionFn"
  [identityFn]="identityFn">
  </adr-autocomplete>`,
});

export const reactiveForm = reactiveFormTemplate.bind({});
reactiveForm.args = {
    options: [
        { id: 1, firstName: 'Gaetan', name: 'Redin' },
        { id: 2, firstName: 'Soren', name: 'Redin' },
        { id: 3, firstName: 'Lord', name: 'Voldemor' },
    ],
};

const addOptionTemplate: Story<AutocompleteComponent<Person>> = (
    args: AutocompleteComponent<Person>,
) => ({
    moduleMetadata: {
        imports: [AutocompleteModule, ReactiveFormsModule],
    },
    props: {
        ...args,
        valueChange: action('valueChange'),
        value: null,
        displayOptionFn: (option: Person): string => `${option.firstName} - ${option.name}`,
        control: new FormControl({ id: 2, firstName: 'Soren', name: 'Redin' }),
        identityFn: (p: Person): any => p.id,
    },
    template: `
<adr-autocomplete 
[formControl]="control" [openOn]="openOn" [options]="options" 
[required]="required" (valueChange)="valueChange($event)"
[displayOptionFn]="displayOptionFn" [createOptionFn]="createOptionFn"
[identityFn]="identityFn"></adr-autocomplete>`,
});

export const addOption = addOptionTemplate.bind({});
addOption.args = {
    options: [
        { id: 1, firstName: 'Gaetan', name: 'Redin' },
        { id: 2, firstName: 'Soren', name: 'Redin' },
        { id: 3, firstName: 'Lord', name: 'Voldemor' },
    ],
    createOptionFn: function (input: string) {
        if (input.includes('-')) {
            return {
                firstName: input.split('-')[0].trim(),
                name: input.split('-')[1].trim(),
            } as Person;
        }
        return {
            firstName: input,
            name: '',
        } as Person;
    },
};
