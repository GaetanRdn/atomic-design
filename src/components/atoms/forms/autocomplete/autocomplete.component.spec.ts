import { Component, DebugElement } from "@angular/core";
import { TestBed } from "@angular/core/testing";
import { TemplateLookup } from "src/components/core/tests/template-lookup";
import { AutocompleteComponent } from "./autocomplete.component";
import { AutocompleteModule } from "./autocomplete.module";

describe("AutocompleteComponent", () => {
  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [
        HostComponent,
        RequiredHostComponent,
        ObjectValueHostComponent,
        OpenOnInputHostComponent,
      ],
      imports: [AutocompleteModule],
    });
  });

  describe("Basic", () => {
    let templateLookup: TemplateLookup<HostComponent>;

    beforeEach(() => {
      templateLookup = new TemplateLookup<HostComponent>(
        TestBed.createComponent(HostComponent)
      );

      templateLookup.detectChanges();
    });

    test("should create", () => {
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    test("when focus then options panel is open", () => {
      // WHEN
      templateLookup.query("input").focus();
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    test("when click on option then panel is closed", () => {
      // GIVEN
      templateLookup.query("input").focus();
      templateLookup.detectChanges();

      // WHEN
      templateLookup.query(".adr-option:last-child").click();
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    test("when click on option then check value", () => {
      // GIVEN
      templateLookup.query("input").focus();
      templateLookup.detectChanges();

      // WHEN
      templateLookup.query(".adr-option").click();
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.getComponent(AutocompleteComponent).value).toEqual(
        "Gaetan"
      );
      expect(templateLookup.hostComponent.value).toEqual("Gaetan");
    });

    test("when click on selected option then value is null", () => {
      // GIVEN
      templateLookup.query("input").focus();
      templateLookup.detectChanges();

      // WHEN
      templateLookup.query(".adr-option.adr-selected").click();
      templateLookup.detectChanges();

      // THEN
      expect(
        templateLookup.getComponent(AutocompleteComponent).value
      ).toBeNull();
      expect(templateLookup.hostComponent.value).toBeNull();
    });

    test("when click outside then options panel is closed", () => {
      // GIVEN
      templateLookup.query("input").focus();
      templateLookup.detectChanges();

      // WHEN
      templateLookup.query(".outside").click();
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    test("when input value then displayed options are filtered", () => {
      // GIVEN
      const input: DebugElement = templateLookup.get("input");
      input.nativeElement.value = "g";
      input.triggerEventHandler("input", { target: input.nativeElement });

      // WHEN
      templateLookup.query("input").focus();
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });
  });

  describe("Required", () => {
    let templateLookup: TemplateLookup<RequiredHostComponent>;

    beforeEach(() => {
      templateLookup = new TemplateLookup<RequiredHostComponent>(
        TestBed.createComponent(RequiredHostComponent)
      );

      templateLookup.detectChanges();
    });

    test("should create", () => {
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    test("when required and click on selected option then value does not change", () => {
      // GIVEN
      jest.spyOn(
        templateLookup.getComponent(AutocompleteComponent).valueChange,
        "emit"
      );
      templateLookup.query("input").focus();
      templateLookup.detectChanges();

      // WHEN
      templateLookup.query(".adr-option.adr-selected").click();
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
      expect(
        templateLookup.getComponent(AutocompleteComponent).valueChange.emit
      ).not.toBeCalled();
      expect(templateLookup.getComponent(AutocompleteComponent).value).toEqual(
        "Soren"
      );
      expect(templateLookup.hostComponent.value).toEqual("Soren");
    });
  });

  describe("Objects as values", () => {
    let templateLookup: TemplateLookup<ObjectValueHostComponent>;

    beforeEach(() => {
      templateLookup = new TemplateLookup<ObjectValueHostComponent>(
        TestBed.createComponent(ObjectValueHostComponent)
      );

      templateLookup.detectChanges();
    });

    test("should create", () => {
      // WHEN
      templateLookup.query("input").focus();
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });
  });

  describe("Open on input", () => {
    let templateLookup: TemplateLookup<OpenOnInputHostComponent>;

    beforeEach(() => {
      templateLookup = new TemplateLookup<OpenOnInputHostComponent>(
        TestBed.createComponent(OpenOnInputHostComponent)
      );

      templateLookup.detectChanges();
    });

    test("should not open on focus", () => {
      // WHEN
      templateLookup.query("input").focus();
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    test("should open on input with value", () => {
      // GIVEN
      const input: DebugElement = templateLookup.get("input");
      input.nativeElement.value = "s";
      templateLookup.detectChanges();

      // WHEN
      input.triggerEventHandler("input", { target: input.nativeElement });
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });
  });
});

@Component({
  template: ` <adr-autocomplete
      [options]="options"
      [(value)]="value"
    ></adr-autocomplete>
    <span class="outside"></span>`,
})
class HostComponent {
  public options: string[] = ["Soren", "Gaetan"];

  public value: string = "Soren";
}

@Component({
  template: ` <adr-autocomplete
    [options]="options"
    [(value)]="value"
    required
  ></adr-autocomplete>`,
})
class RequiredHostComponent extends HostComponent {}

interface Person {
  id: number;
  firstName: string;
}

@Component({
  template: ` <adr-autocomplete
    [options]="options"
    [(value)]="value"
    openOn="input"
  ></adr-autocomplete>`,
})
class OpenOnInputHostComponent extends HostComponent {}

interface Person {
  id: number;
  firstName: string;
}

@Component({
  template: ` <adr-autocomplete
      [options]="options"
      [(value)]="value"
      [displayOptionFn]="displayFn"
      required
    ></adr-autocomplete>
    <span class="outside"></span>`,
})
class ObjectValueHostComponent {
  public options: Person[] = [
    { id: 1, firstName: "Soren" },
    { id: 2, firstName: "Gaetan" },
  ];

  public value: Person = { id: 1, firstName: "Soren" };

  public displayFn: (option: Person) => string = (option: Person): string =>
    option.firstName;
}
