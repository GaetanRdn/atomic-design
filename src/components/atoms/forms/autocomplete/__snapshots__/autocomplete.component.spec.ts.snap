// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutocompleteComponent Basic should create 1`] = `
<adr-autocomplete>
  <input
    adrinput=""
    class="adr-input"
    value="Soren"
  />
  
</adr-autocomplete>
`;

exports[`AutocompleteComponent Basic when click on option then panel is closed 1`] = `
<adr-autocomplete>
  <input
    adrinput=""
    class="adr-input adr-focused"
    value=""
  />
  
</adr-autocomplete>
`;

exports[`AutocompleteComponent Basic when click outside then options panel is closed 1`] = `
<adr-autocomplete>
  <input
    adrinput=""
    class="adr-input adr-focused"
    value="Soren"
  />
  
</adr-autocomplete>
`;

exports[`AutocompleteComponent Basic when focus then options panel is open 1`] = `
<adr-autocomplete
  class="adr-opened"
>
  <input
    adrinput=""
    class="adr-input adr-focused"
    value="Soren"
  />
  
  <div
    class="adr-autocomplete-options"
  >
    
    <span
      class="adr-option"
    >
      Gaetan
    </span>
    <span
      class="adr-option adr-selected"
    >
      Soren
    </span>
  </div>
</adr-autocomplete>
`;

exports[`AutocompleteComponent Basic when input value then displayed options are filtered 1`] = `
<adr-autocomplete
  class="adr-opened"
>
  <input
    adrinput=""
    class="adr-input adr-focused"
    value="g"
  />
  
  <div
    class="adr-autocomplete-options"
  >
    
    <span
      class="adr-option"
    >
      Gaetan
    </span>
  </div>
</adr-autocomplete>
`;

exports[`AutocompleteComponent Objects as values should create 1`] = `
<adr-autocomplete
  class="adr-opened"
  required="true"
>
  <input
    adrinput=""
    class="adr-input adr-focused"
    value="Soren"
  />
  
  <div
    class="adr-autocomplete-options"
  >
    
    <span
      class="adr-option"
    >
      Soren
    </span>
    <span
      class="adr-option"
    >
      Gaetan
    </span>
  </div>
</adr-autocomplete>
`;

exports[`AutocompleteComponent Open on input should not open on focus 1`] = `
<adr-autocomplete
  openon="input"
>
  <input
    adrinput=""
    class="adr-input adr-focused"
    value="Soren"
  />
  
</adr-autocomplete>
`;

exports[`AutocompleteComponent Open on input should open on input with value 1`] = `
<adr-autocomplete
  class="adr-opened"
  openon="input"
>
  <input
    adrinput=""
    class="adr-input"
    value="s"
  />
  
  <div
    class="adr-autocomplete-options"
  >
    
    <span
      class="adr-option adr-selected"
    >
      Soren
    </span>
  </div>
</adr-autocomplete>
`;

exports[`AutocompleteComponent Required should create 1`] = `
<adr-autocomplete
  required="true"
>
  <input
    adrinput=""
    class="adr-input"
    value="Soren"
  />
  
</adr-autocomplete>
`;

exports[`AutocompleteComponent Required when required and click on selected option then value does not change 1`] = `
<adr-autocomplete
  class="adr-opened"
  required="true"
>
  <input
    adrinput=""
    class="adr-input adr-focused"
    value="Soren"
  />
  
  <div
    class="adr-autocomplete-options"
  >
    
    <span
      class="adr-option"
    >
      Gaetan
    </span>
    <span
      class="adr-option adr-selected"
    >
      Soren
    </span>
  </div>
</adr-autocomplete>
`;
